### PR TITLE
AutoRegisteringObjectGraphType improvement

### LINKED ISSUE #1079: AutoRegisteringObjectGraphType Improvement
## Summary

It would be nice to have the support of the following features in `AutoRegisteringObjectGraphType<T>`:
1. Enum properties support.
2. Support for property types registered in `GraphTypeTypeRegistry`.
3. Support for `RequiredAttribute` to tune nullability of reference typed properties.

### Motivation

Simplification of the schema.

### PR DESCRIPTION
**Pull request by sungam3r**
Fixes #1079

### CODE REVIEW

File: src/GraphQL.Tests/Types/ComplexGraphTypeTests.cs [hunk 1]
```diff
@@ -91,7 +138,7 @@ public void infers_from_list_types()
 
             type.Field(d => d.someList, nullable: true);
 
-            type.Fields.Last().Type.ShouldBe(typeof(ListGraphType<IntGraphType>));
+            type.Fields.Last().Type.ShouldBe(typeof(ListGraphType<NonNullGraphType<IntGraphType>>));
```
Comment (sungam3r) [inline src/GraphQL.Tests/Types/ComplexGraphTypeTests.cs]: It seems to me that the behavior regarding this property was wrong:
```c#
 public List<int> someList { get; set; }
```
Elements of this list can never have null value, so it should be coerced to `ListGraphType<NonNullGraphType<IntGraphType>>`.

I added tests for various combinations of types.

File: src/GraphQL/GraphQL.csproj [hunk 2]
```diff
@@ -29,6 +29,7 @@
     <PackageReference Include="GraphQL-Parser" Version="3.0.0" />
     <PackageReference Include="Newtonsoft.Json" Version="10.0.3" />
     <PackageReference Include="System.Buffers" Version="4.5.0" />
+    <PackageReference Include="System.ComponentModel.Annotations" Version="4.5.0" />
```
Comment (sungam3r) [inline src/GraphQL/GraphQL.csproj]: Is it permissible to add this dependency?

File: src/GraphQL/TypeExtensions.cs [hunk 3]
```diff
@@ -129,20 +129,23 @@ public static Type GetGraphTypeFromType(this Type type, bool isNullable = false)
...
-            if (IsAnIEnumerable(type))
+            else if (IsAnIEnumerable(type))
+            {
+                var clrElementType = GetEnumerableElementType(type);
+                var elementType = GetGraphTypeFromType(clrElementType, clrElementType.IsNullable()); // isNullable from elementType, not from parent container
+                graphType = typeof(ListGraphType<>).MakeGenericType(elementType);
+            }
+            else
             {
-                var elementType = GetGraphTypeFromType(type.GenericTypeArguments.First(), isNullable);
```
Comment (sungam3r) [inline src/GraphQL/TypeExtensions.cs]: There was an error in the case of `IEnumerable` property (not `IEnumerable<T>`) - `GenericTypeArguments` returns empty array.

File: src/GraphQL/Types/AutoRegisteringObjectGraphType.cs [hunk 4]
```diff
@@ -48,7 +61,45 @@ protected virtual IEnumerable<PropertyInfo> GetRegisteredProperties()
...
                 .GetProperties(BindingFlags.Public | BindingFlags.Instance)
-                .Where(p => p.PropertyType.IsValueType || p.PropertyType == typeof(string));
+                .Where(p => IsEnabledForRegister(p.PropertyType, true));
+        }
+
+        private static bool IsEnabledForRegister(Type propertyType, bool firstCall)
+        {
+            if (propertyType == typeof(string)) return true;
+
+            if (propertyType.IsValueType) return true; // TODO: requires discussion: Nullable<T>, enums, any struct
```
Comment (sungam3r) [inline src/GraphQL/Types/AutoRegisteringObjectGraphType.cs]: This behavior seems to me to be wrong. You can’t just take and allow to register all structures/enums.
While left as it was before.

File: src/GraphQL/Utilities/GraphTypeTypeRegistry.cs [hunk 5]
```diff
@@ -22,33 +22,28 @@ static GraphTypeTypeRegistry()
                 [typeof(DateTime)] = typeof(DateGraphType),
                 [typeof(DateTimeOffset)] = typeof(DateTimeOffsetGraphType),
                 [typeof(TimeSpan)] = typeof(TimeSpanSecondsGraphType),
-                [typeof(Guid)] = typeof(IdGraphType)
+                [typeof(Guid)] = typeof(IdGraphType),
+                [typeof(short)] = typeof(ShortGraphType),
```
Comment (sungam3r) [inline src/GraphQL/Utilities/GraphTypeTypeRegistry.cs]: And again forgotten scalars ...

File: src/GraphQL/GraphQL.csproj (hunk 2 above)
Comment (joemcbride) [reply src/GraphQL/GraphQL.csproj]: Admittedly it makes me cringe a bit but that is probably the easiest solution.

Comment (sungam3r) [reply src/GraphQL/GraphQL.csproj]: Well, this is essentially standardized well-known library with common meta-information. If not this, then what? I do not really want to reinvent the wheel.

**Pull request merged**